File: dynabuffers-python/dynabuffers/ast/datatype/BytearrayType.py
from dynabuffers.api.ISerializable import ISerializable


class BytearrayType(ISerializable):

    def size(self, value, registry):
        return 4 + len(value)

    def serialize(self, value, buffer, registry):
        buffer.putInt(len(value))
        buffer.put(value)

    def deserialize(self, buffer, registry):
        length = buffer.getInt()
        array = buffer.get(length)

        return array

File: dynabuffers-java/src/main/kotlin/dynabuffers/ast/datatype/BytearrayType.kt
package dynabuffers.ast.datatype

import dynabuffers.api.IRegistry
import dynabuffers.api.ISerializable
import dynabuffers.api.IType
import java.nio.ByteBuffer

class BytearrayType : IType, ISerializable {

    override fun size(value: Any, registry: IRegistry) = 4 + (value as ByteArray).size

    override fun serialize(value: Any, buffer: ByteBuffer, registry: IRegistry) {
        val bytes = (value as ByteArray)
        buffer.putInt(bytes.size)
        buffer.put(bytes)
    }

    override fun deserialize(buffer: ByteBuffer, registry: IRegistry): Any {
        val length = buffer.int
        val array = ByteArray(length)
        buffer.get(array)

        return array
    }

}
